sql/api/src/main/scala/org/apache/spark/sql/catalyst/encoders/AgnosticEncoder.scala

```diff
@@ -209,7 +209,8 @@ object AgnosticEncoders {
 
   // Nullable leaf encoders
   case object NullEncoder extends LeafEncoder[java.lang.Void](NullType)
-  case object StringEncoder extends LeafEncoder[String](StringType)
+  class StringEncoder(strType: DataType) extends LeafEncoder[String](strType)
```

Comment:
char/varchar has length parameter, so this can't be a singleton. We can create multiple classes though, but then we need to update all the patter match places for this.